src/java_binding/java/java-binding/src/main/java/com/risingwave/java/binding/rpc/MetaClient.java

```diff
@@ -0,0 +1,136 @@
+package com.risingwave.java.binding.rpc;
+
+import com.risingwave.proto.Catalog.Table;
+import com.risingwave.proto.ClusterServiceGrpc;
+import com.risingwave.proto.ClusterServiceGrpc.ClusterServiceBlockingStub;
+import com.risingwave.proto.Common.HostAddress;
+import com.risingwave.proto.Common.WorkerType;
+import com.risingwave.proto.DdlServiceGrpc;
+import com.risingwave.proto.DdlServiceGrpc.DdlServiceBlockingStub;
+import com.risingwave.proto.DdlServiceOuterClass.JavaGetTableRequest;
+import com.risingwave.proto.DdlServiceOuterClass.JavaGetTableResponse;
+import com.risingwave.proto.HeartbeatServiceGrpc;
+import com.risingwave.proto.HeartbeatServiceGrpc.HeartbeatServiceBlockingStub;
+import com.risingwave.proto.Hummock.HummockVersion;
+import com.risingwave.proto.Hummock.JavaPinVersionRequest;
+import com.risingwave.proto.Hummock.JavaPinVersionResponse;
+import com.risingwave.proto.HummockManagerServiceGrpc;
+import com.risingwave.proto.HummockManagerServiceGrpc.HummockManagerServiceBlockingStub;
+import com.risingwave.proto.Meta.AddWorkerNodeRequest;
+import com.risingwave.proto.Meta.AddWorkerNodeResponse;
+import com.risingwave.proto.Meta.HeartbeatRequest;
+import io.grpc.Grpc;
+import io.grpc.InsecureChannelCredentials;
+import io.grpc.ManagedChannel;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+public class MetaClient implements AutoCloseable {
+    private final int workerId;
+
+    private final ManagedChannel channel;
+
+    // Scheduler for background tasks.
+    private final ScheduledExecutorService scheduler;
+
+    // RPC stubs.
+    private final ClusterServiceBlockingStub clusterStub;
+    private final DdlServiceBlockingStub ddlStub;
+    private final HeartbeatServiceBlockingStub heartbeatStub;
+    private final HummockManagerServiceBlockingStub hummockStub;
+
+    private boolean isClosed;
+
+    // A heart beat task that sends a heartbeat to the meta service when run.
+    private class HeartbeatTask implements Runnable {
+        Instant lastHeartbeatSent;
+        Duration timeout;
+
+        HeartbeatTask(Duration timeout) {
+            this.lastHeartbeatSent = Instant.now();
+            this.timeout = timeout;
+        }
+
+        @Override
+        public void run() {
+            HeartbeatRequest req = HeartbeatRequest.newBuilder().setNodeId(workerId).build();
+
+            try {
+                heartbeatStub
+                        .withDeadlineAfter(timeout.toMillis(), TimeUnit.MILLISECONDS)
+                        .heartbeat(req);
+            } catch (Exception e) {
+                Logger.getGlobal().warning(String.format("Failed to send heartbeat: %s", e));
+            }
+
+            Instant now = Instant.now();
+            if (Duration.between(lastHeartbeatSent, Instant.now()).compareTo(timeout) > 0) {
```

Comment:
The heart beat is used to avoid meta node state cleaning but not for the cleaning on the meta client side. Therefore, the meta client should not proactively exit on timeout but may keep retrying the heartbeat with like exponentially increasing retry interval until we gets some error when reading the data and gets terminated externally.

Reply:
I just tried to make the Java implementation as similar as possible to the Rust one, which does not seem to adopt exponential backoff either.

https://github.com/risingwavelabs/risingwave/blob/7b0f2d21d278013faca232d399d0366352b240e0/src/rpc_client/src/meta_client.rs#L424-L481

 Maybe we should fix this in another PR?

Reply:
Make sense. We can implement the exponential backoff in later PR. 

But in this PR, I think we should at least avoid `exit` when timeout happens, because for the current java application, we are simply a reader, and losing the heartbeat is not bound to affecting the read consistency, and therefore we can continue reading the data until the data to read are unfortunately cleaned and report an error. We can simply endlessly retry sending the heartbeat.